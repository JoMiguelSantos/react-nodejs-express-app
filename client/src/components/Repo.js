import React, { useState, useEffect } from "react";
import { useDispatch } from "react-redux";
import { capitalize } from "../utils";
import "./Repo.css";
import { addBookmark, delBookmark } from "../store/actions";

export default ({
  id,
  name,
  html_url,
  description,
  created_at,
  updated_at,
  stargazers_count,
  watchers_count,
  language,
  forks_count,
  open_issues_count,
  forks,
  open_issues,
  isRepoBookmarked
}) => {
  const [isBookmarked, setBookmarked] = useState(false);
  const dispatch = useDispatch();

  useEffect(() => {
    isRepoBookmarked && setBookmarked(true);
  }, [isRepoBookmarked]);

  const clickHandler = () => {
    setBookmarked(!isBookmarked);
    isBookmarked ? dispatch(delBookmark(id)) : dispatch(addBookmark(id));
  };

  const repoDescriptors = {
    id: id,
    name: name,
    html_url: html_url,
    description: description,
    created_at: created_at,
    updated_at: updated_at,
    stargazers_count: stargazers_count,
    watchers_count: watchers_count,
    language: language,
    open_issues_count: open_issues_count,
    forks: forks,
    open_issues: open_issues
  };

  return (
    <div className="repo__descriptor--items">
      {Object.entries(repoDescriptors)
        .slice(1)
        .map(([key, value]) => {
          return (
            <div key={key} className="repo__descriptor--item">
              <label>{`${capitalize(key, "_")}:`}</label>
              <p>{value}</p>
            </div>
          );
        })}
      <button onClick={clickHandler}>
        {isBookmarked ? "Unbookmark" : "Bookmark"}
      </button>
    </div>
  );
};
