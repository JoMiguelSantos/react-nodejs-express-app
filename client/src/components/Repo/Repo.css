.repo__descriptor--items {
  display: flex;
  align-items: center;
  justify-content: flex-start;
  flex-wrap: wrap;
  /* align-items: flex-start; */
  padding: 1rem;
  /* margin: 0; */
  width: 100%;
  border-bottom: 3px ridge blue;
}

.repo__descriptor--item {
  display: flex;
  align-items: center;
  justify-content: flex-start;
  padding: 0 1rem;
}

.repo__descriptor--items p {
  font-size: 0.8rem;
  padding: 0.2rem;
  margin: 0;
}

.repo__descriptor--items label {
  color: darkblue;
  font-size: 0.8rem;
  padding: 0.2rem;
  margin: 0;
}

.btn-submit {
  align-self: end;
  border-radius: 20px;
  background-color: azure;
  font-weight: 700;
  margin-left: auto;
}
