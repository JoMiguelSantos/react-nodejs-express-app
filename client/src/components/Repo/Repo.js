import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { capitalize } from "../../utils";
import "./Repo.css";
import { addBookmark, delBookmark } from "../../store/actions";

export default ({
  id,
  name,
  html_url,
  description,
  created_at,
  updated_at,
  stargazers_count,
  watchers_count,
  language,
  open_issues_count,
  forks_count
}) => {
  const dispatch = useDispatch();
  const isBookmarked = useSelector(
    state =>
      state.bookmarks.bookmarks &&
      !!state.bookmarks.bookmarks.find(repo => repo.id === id)
  );

  const repoDescriptors = {
    id: id,
    name: name,
    html_url: html_url,
    description: description,
    created_at: created_at,
    updated_at: updated_at,
    stargazers_count: stargazers_count,
    watchers_count: watchers_count,
    language: language,
    open_issues_count: open_issues_count,
    forks_count: forks_count
  };

  const clickHandler = async () => {
    if (isBookmarked) {
      const data = await fetch(`http://localhost:4000/api/v1/bookmarks`, {
        method: "DELETE",
        headers: {
          "Content-Type": "application/json"
        },
        body: JSON.stringify({ repoId: id })
      });
      data.status === 204 && dispatch(delBookmark(id));
    } else {
      const data = await fetch(`http://localhost:4000/api/v1/bookmarks`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json"
        },
        body: JSON.stringify({ repoId: id })
      });
      data.status === 201 && dispatch(addBookmark(repoDescriptors));
    }
  };

  return (
    <div className="repo__descriptor--items">
      {Object.entries(repoDescriptors)
        .slice(1)
        .map(([key, value]) => {
          return (
            <div key={key} className="repo__descriptor--item">
              <label>{`${capitalize(key, "_")}:`}</label>
              <p>{value}</p>
            </div>
          );
        })}
      <button
        className={`btn-bookmark ${isBookmarked && `active`}`}
        onClick={clickHandler}
      >
        {isBookmarked ? "Unbookmark" : "Bookmark"}
      </button>
    </div>
  );
};
