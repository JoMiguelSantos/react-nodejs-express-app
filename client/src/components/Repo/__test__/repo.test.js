import Repo from "../Repo";
import React from "react";
import { createStore, combineReducers } from "redux";
import { Provider } from "react-redux";
import { render, waitForElement, fireEvent } from "@testing-library/react";
import reducer, { initialState } from "../../../store/reducers/bookmarks";
import { FetchMock } from "@react-mock/fetch";

// need to add composeReducers if want to test reducers/actions
function renderWithReduxFetch({
  initialState,
  store = createStore(combineReducers({ bookmarks: reducer }), initialState)
} = {}) {
  return {
    ...render(
      <FetchMock
        mocks={[
          {
            matcher: "http://localhost:4000/api/v1/bookmarks",
            method: "DELETE",
            response: 204
          },
          {
            matcher: "http://localhost:4000/api/v1/bookmarks",
            method: "POST",
            response: 201
          }
        ]}
      >
        <Provider store={store}>
          <Repo />
        </Provider>
      </FetchMock>
    ),
    store
  };
}

// test renders correct Repo item given data example
const mockRepo = {
  id: 183743694,
  name: "node-express-app",
  html_url: "https://github.com/node-express-app",
  description: "This a repo made for an interview challenge",
  created_at: "2020-02-25T12:23:12",
  updated_at: "2020-02-25T16:23:12",
  stargazers_count: 14554,
  watchers_count: 3563,
  language: "javascript",
  open_issues_count: 124,
  forks_count: 1355
};

describe("<Repo /> unit tests", () => {
  // SNAPSHOT
  it("matches the previous snapshot", () => {
    const { container } = renderWithReduxFetch();
    expect(container).toMatchSnapshot();
  });

  it("renders initial state with Bookmark text visible", async () => {
    const { getByText } = renderWithReduxFetch();
    await waitForElement(() => getByText("Bookmark"));
  });

  it("toggles un/bookmark", async () => {
    const { getByText } = renderWithReduxFetch();
    await waitForElement(() => getByText("Bookmark"));

    fireEvent.click(getByText("Bookmark"));
    await waitForElement(() => getByText("Unbookmark"));

    fireEvent.click(getByText("Unbookmark"));
    await waitForElement(() => getByText("Bookmark"));
  });
});
