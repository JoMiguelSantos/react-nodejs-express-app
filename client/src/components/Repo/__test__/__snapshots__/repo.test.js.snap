// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Repo /> unit tests matches the previous snapshot 1`] = `
<div>
  <div
    class="repo__descriptor--items"
  >
    <div
      class="repo__descriptor--item"
    >
      <label>
        Name:
      </label>
      <p />
    </div>
    <div
      class="repo__descriptor--item"
    >
      <label>
        Html Url:
      </label>
      <p />
    </div>
    <div
      class="repo__descriptor--item"
    >
      <label>
        Description:
      </label>
      <p />
    </div>
    <div
      class="repo__descriptor--item"
    >
      <label>
        Created At:
      </label>
      <p />
    </div>
    <div
      class="repo__descriptor--item"
    >
      <label>
        Updated At:
      </label>
      <p />
    </div>
    <div
      class="repo__descriptor--item"
    >
      <label>
        Stargazers Count:
      </label>
      <p />
    </div>
    <div
      class="repo__descriptor--item"
    >
      <label>
        Watchers Count:
      </label>
      <p />
    </div>
    <div
      class="repo__descriptor--item"
    >
      <label>
        Language:
      </label>
      <p />
    </div>
    <div
      class="repo__descriptor--item"
    >
      <label>
        Open Issues Count:
      </label>
      <p />
    </div>
    <div
      class="repo__descriptor--item"
    >
      <label>
        Forks Count:
      </label>
      <p />
    </div>
    <button
      class="btn-bookmark false"
    >
      Bookmark
    </button>
  </div>
</div>
`;
